File: src/compiler/prepare_grammar/extract_simple_aliases.cc
#include "compiler/prepare_grammar/extract_simple_aliases.h"
#include "compiler/lexical_grammar.h"
#include "compiler/syntax_grammar.h"
#include <unordered_map>
#include <vector>

namespace tree_sitter {
namespace prepare_grammar {

using std::pair;
using std::vector;
using std::unordered_map;
using rules::Alias;
using rules::Symbol;

template <typename T>
static void apply_alias(T *variable, Alias alias) {
  if (!alias.value.empty()) {
    variable->name = alias.value;
    variable->type = alias.is_named ? VariableTypeNamed : VariableTypeAnonymous;
  }
}

std::unordered_map<rules::Symbol, rules::Alias>
extract_simple_aliases(SyntaxGrammar *syntax_grammar, LexicalGrammar *lexical_grammar) {
  struct SymbolStatus {
    Alias alias;
    bool eligible = true;
  };

  vector<SymbolStatus> terminal_status_list(lexical_grammar->variables.size());
  vector<SymbolStatus> non_terminal_status_list(syntax_grammar->variables.size());
  vector<SymbolStatus> external_status_list(syntax_grammar->external_tokens.size());

  for (const SyntaxVariable &variable : syntax_grammar->variables) {
    for (const Production &production : variable.productions) {
      for (const ProductionStep &step : production.steps) {
        SymbolStatus *status;
        if (step.symbol.is_built_in())  {
          continue;
        } else if (step.symbol.is_external()) {
          status = &external_status_list[step.symbol.index];
        } else if (step.symbol.is_terminal()) {
          status = &terminal_status_list[step.symbol.index];
        } else {
          status = &non_terminal_status_list[step.symbol.index];
        }

        if (step.alias.value.empty()) {
          status->alias = Alias();
          status->eligible = false;
        }

        if (status->eligible) {
          if (status->alias.value.empty()) {
            status->alias = step.alias;
          } else if (status->alias != step.alias) {
            status->alias = Alias();
            status->eligible = false;
          }
        }
      }
    }
  }

  for (SyntaxVariable &variable : syntax_grammar->variables) {
    for (Production &production : variable.productions) {
      for (ProductionStep &step : production.steps) {
        SymbolStatus *status;
        if (step.symbol.is_built_in())  {
          continue;
        } else if (step.symbol.is_external()) {
          status = &external_status_list[step.symbol.index];
        } else if (step.symbol.is_terminal()) {
          status = &terminal_status_list[step.symbol.index];
        } else {
          status = &non_terminal_status_list[step.symbol.index];
        }

        if (!status->alias.value.empty()) {
          step.alias = Alias();
        }
      }
    }
  }

  unordered_map<Symbol, Alias> result;

  for (unsigned i = 0, n = terminal_status_list.size(); i < n; i++) {
    if (!terminal_status_list[i].alias.value.empty()) {
      result[Symbol::terminal(i)] = terminal_status_list[i].alias;
    }
  }

  for (unsigned i = 0, n = non_terminal_status_list.size(); i < n; i++) {
    if (!non_terminal_status_list[i].alias.value.empty()) {
      result[Symbol::non_terminal(i)] = non_terminal_status_list[i].alias;
    }
  }

  for (unsigned i = 0, n = external_status_list.size(); i < n; i++) {
    if (!external_status_list[i].alias.value.empty()) {
      result[Symbol::external(i)] = external_status_list[i].alias;
    }
  }

  return result;
}

}  // namespace prepare_grammar
}  // namespace tree_sitter

File: cli/src/generate/prepare_grammar/extract_simple_aliases.rs
use crate::generate::grammars::{LexicalGrammar, SyntaxGrammar};
use crate::generate::rules::{Alias, AliasMap, Symbol, SymbolType};

#[derive(Clone, Default)]
struct SymbolStatus {
    alias: Option<Alias>,
    conflicting: bool,
}

pub(super) fn extract_simple_aliases(
    syntax_grammar: &mut SyntaxGrammar,
    lexical_grammar: &LexicalGrammar,
) -> AliasMap {
    // Determine which symbols in the grammars are *always* aliased to a single name.
    let mut terminal_status_list = vec![SymbolStatus::default(); lexical_grammar.variables.len()];
    let mut non_terminal_status_list =
        vec![SymbolStatus::default(); syntax_grammar.variables.len()];
    let mut external_status_list =
        vec![SymbolStatus::default(); syntax_grammar.external_tokens.len()];
    for variable in syntax_grammar.variables.iter() {
        for production in variable.productions.iter() {
            for step in production.steps.iter() {
                let mut status = match step.symbol {
                    Symbol {
                        kind: SymbolType::External,
                        index,
                    } => &mut external_status_list[index],
                    Symbol {
                        kind: SymbolType::NonTerminal,
                        index,
                    } => &mut non_terminal_status_list[index],
                    Symbol {
                        kind: SymbolType::Terminal,
                        index,
                    } => &mut terminal_status_list[index],
                    Symbol {
                        kind: SymbolType::End,
                        ..
                    } => panic!("Unexpected end token"),
                };

                if step.alias.is_none() {
                    status.alias = None;
                    status.conflicting = true;
                }

                if !status.conflicting {
                    if status.alias.is_none() {
                        status.alias = step.alias.clone();
                    } else if status.alias != step.alias {
                        status.alias = None;
                        status.conflicting = true;
                    }
                }
            }
        }
    }

    // Remove the aliases for those symbols.
    for variable in syntax_grammar.variables.iter_mut() {
        for production in variable.productions.iter_mut() {
            for step in production.steps.iter_mut() {
                let status = match step.symbol {
                    Symbol {
                        kind: SymbolType::External,
                        index,
                    } => &external_status_list[index],
                    Symbol {
                        kind: SymbolType::NonTerminal,
                        index,
                    } => &non_terminal_status_list[index],
                    Symbol {
                        kind: SymbolType::Terminal,
                        index,
                    } => &terminal_status_list[index],
                    Symbol {
                        kind: SymbolType::End,
                        ..
                    } => panic!("Unexpected end token"),
                };

                if status.alias.is_some() {
                    step.alias = None;
                }
            }
        }
    }

    // Populate a map of the symbols to their aliases.
    let mut result = AliasMap::new();
    for (i, status) in terminal_status_list.into_iter().enumerate() {
        if let Some(alias) = status.alias {
            result.insert(Symbol::terminal(i), alias);
        }
    }
    for (i, status) in non_terminal_status_list.into_iter().enumerate() {
        if let Some(alias) = status.alias {
            result.insert(Symbol::non_terminal(i), alias);
        }
    }
    for (i, status) in external_status_list.into_iter().enumerate() {
        if let Some(alias) = status.alias {
            result.insert(Symbol::external(i), alias);
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::generate::grammars::{
        LexicalVariable, Production, ProductionStep, SyntaxVariable, VariableType,
    };
    use crate::generate::nfa::Nfa;

    #[test]
    fn test_extract_simple_aliases() {
        let mut syntax_grammar = SyntaxGrammar {
            variables: vec![
                SyntaxVariable {
                    name: "v1".to_owned(),
                    kind: VariableType::Named,
                    productions: vec![Production {
                        dynamic_precedence: 0,
                        steps: vec![
                            ProductionStep::new(Symbol::terminal(0)).with_alias("a1", true),
                            ProductionStep::new(Symbol::terminal(1)).with_alias("a2", true),
                            ProductionStep::new(Symbol::terminal(2)).with_alias("a3", true),
                        ],
                    }],
                },
                SyntaxVariable {
                    name: "v2".to_owned(),
                    kind: VariableType::Named,
                    productions: vec![Production {
                        dynamic_precedence: 0,
                        steps: vec![
                            // Token 0 is always aliased as "a1".
                            ProductionStep::new(Symbol::terminal(0)).with_alias("a1", true),
                            // Token 1 is aliased above, but not here.
                            ProductionStep::new(Symbol::terminal(1)),
                            // Token 2 is aliased differently than above.
                            ProductionStep::new(Symbol::terminal(2)).with_alias("a4", true),
                        ],
                    }],
                },
            ],
            extra_symbols: Vec::new(),
            expected_conflicts: Vec::new(),
            variables_to_inline: Vec::new(),
            supertype_symbols: Vec::new(),
            external_tokens: Vec::new(),
            word_token: None,
        };

        let lexical_grammar = LexicalGrammar {
            nfa: Nfa::new(),
            variables: vec![
                LexicalVariable {
                    name: "t1".to_string(),
                    kind: VariableType::Anonymous,
                    implicit_precedence: 0,
                    start_state: 0,
                },
                LexicalVariable {
                    name: "t2".to_string(),
                    kind: VariableType::Anonymous,
                    implicit_precedence: 0,
                    start_state: 0,
                },
                LexicalVariable {
                    name: "t3".to_string(),
                    kind: VariableType::Anonymous,
                    implicit_precedence: 0,
                    start_state: 0,
                },
            ],
        };

        let simple_aliases = extract_simple_aliases(&mut syntax_grammar, &lexical_grammar);
        assert_eq!(simple_aliases.len(), 1);
        assert_eq!(
            simple_aliases[&Symbol::terminal(0)],
            Alias {
                value: "a1".to_string(),
                is_named: true,
            }
        );

        assert_eq!(
            syntax_grammar.variables,
            vec![
                SyntaxVariable {
                    name: "v1".to_owned(),
                    kind: VariableType::Named,
                    productions: vec![Production {
                        dynamic_precedence: 0,
                        steps: vec![
                            // 'Simple' alias removed
                            ProductionStep::new(Symbol::terminal(0)),
                            // Other aliases unchanged
                            ProductionStep::new(Symbol::terminal(1)).with_alias("a2", true),
                            ProductionStep::new(Symbol::terminal(2)).with_alias("a3", true),
                        ],
                    },],
                },
                SyntaxVariable {
                    name: "v2".to_owned(),
                    kind: VariableType::Named,
                    productions: vec![Production {
                        dynamic_precedence: 0,
                        steps: vec![
                            ProductionStep::new(Symbol::terminal(0)),
                            ProductionStep::new(Symbol::terminal(1)),
                            ProductionStep::new(Symbol::terminal(2)).with_alias("a4", true),
                        ],
                    },],
                },
            ]
        );
    }
}
